core/templates/dev/head/galleries/Gallery.js

```diff
@@ -203,6 +203,125 @@ oppia.controller('Gallery', [
       $rootScope.loadingMessage = '';
     });
 
+    // Below is the width of each tile
+    var TILEDISPLAYWIDTH = 211.953;
+    // If the value below is changed, the corresponding CSS values must be
+    // changed in oppia.css.
+    var MAXTILEDISPLAY = 4;
+    var maxGalleryTileWidth = TILEDISPLAYWIDTH * MAXTILEDISPLAY;
+    $scope.tileDisplayCount = 0;
+    $scope.galleryTilesResizedWidth = 0;
+
+    var calculateCarouselCssWidth = function() {
+      var windowWidth = $(window).width() * 0.85;
+      $scope.tileDisplayCount = Math.floor(windowWidth / TILEDISPLAYWIDTH);
+      $scope.galleryTilesResizedWidth = $scope.tileDisplayCount *
+        TILEDISPLAYWIDTH;
+      if ($scope.galleryTilesResizedWidth > maxGalleryTileWidth) {
+        $scope.galleryTilesResizedWidth = maxGalleryTileWidth;
+        $scope.tileDisplayCount = MAXTILEDISPLAY;
+      }
+      $('.oppia-gallery-tiles-carousel').css({
+        width: $scope.galleryTilesResizedWidth + 'px'
+      });
+
+      // The following determines whether to enable left scroll after resize
+      for (i = 0; i < $('.oppia-gallery-tiles-carousel').length; i++) {
+        var carouselJQuerySelect = '.oppia-gallery-tiles:eq(n)'.replace(
+          'n', i);
+        var carouselScroll = $(carouselJQuerySelect).scrollLeft();
+        var index = Math.ceil(carouselScroll / TILEDISPLAYWIDTH);
+        $scope.rowIndices[i] = index;
+      }
+    };
+
+    $scope.scroll = function(ind, isLeftScroll) {
+      var carouselJQuerySelect = '.oppia-gallery-tiles:eq(n)'.replace('n', ind);
+      var direction = isLeftScroll ? -1 : 1;
+      var currentScrollPosition = $(carouselJQuerySelect).scrollLeft();
+
+      // Scrolling eligibility checks
+      if ($scope.galleryGroups[ind].activity_summary_dicts.length <=
+          $scope.tileDisplayCount) {
+        return;
+      } else if (isLeftScroll && $scope.rowIndices[ind] == 0) {
+        return;
+      } else if (!isLeftScroll && $scope.rowIndices[ind] >=
+                 $scope.galleryGroups[ind].activity_summary_dicts.length -
+                 $scope.tileDisplayCount) {
+        return;
+      }
+
+      if (currentScrollPosition < 0) {
+        currentScrollPosition = 0;
+      }
+
+      if (isLeftScroll) {
+        $scope.rowIndices[ind] -= $scope.tileDisplayCount;
```

Comment:
Optional: If you like, this could be

    $scope.rowIndices[ind] = Math.max(
      0, $scope.rowIndicesind] - $scope.tileDisplayCount);

and similarly below.